Sources/Common/Communication/EventBus.swift

```diff
@@ -0,0 +1,106 @@
+/// This actor manages the distribution of events to registered observers and uses
+/// `NotificationCenter` for event delivery. It ensures that event handling is thread-safe
+/// and observers are managed efficiently.
+// sourcery: InjectRegisterShared = "EventBus"
+// sourcery: InjectSingleton
+actor SharedEventBus: EventBus {
+    private var notificationCenter: NotificationCenter = .default
+    private var observers: [String: [NSObjectProtocol]] = [:]
+
+    deinit {
+        // Clean up by removing all observers when the EventBus is deinitialized.
+        // Capture a weak reference to self to avoid retain cycles
+        Task { [weak self] in
+            await self?.removeAllObservers()
+        }
+    }
+
+    init() {
+        DIGraphShared.shared.logger.debug("SharedEventBus initialized")
+    }
+
+    /// Posts an event to all registered observers of its type.
+    ///
+    /// - Parameter event: The event to be posted.
+    /// - Returns: True if the event has been posted to any observers, false otherwise.
+    @discardableResult
+    func post(_ event: AnyEventRepresentable) async -> Bool {
+        let key = event.key
+        if let observerList = observers[key], !observerList.isEmpty {
+            notificationCenter.post(name: NSNotification.Name(key), object: event)
+            return true
+        }
+        return false
+    }
+
+    /// Registers an observer for a specific event type.
+    ///
+    /// - Parameters:
+    ///   - eventType: The type of the event to observe.
+    ///   - action: The action to be executed when the event is received.
+    func addObserver(_ eventType: String, action: @escaping (AnyEventRepresentable) -> Void) {
+        let observer = notificationCenter.addObserver(forName: NSNotification.Name(eventType), object: nil, queue: nil) { notification in
+            if let event = notification.object as? AnyEventRepresentable {
+                action(event)
+            }
+        }
+        // Store the observer reference for later management.
+        if observers[eventType] != nil {
+            observers[eventType]?.append(observer)
+        } else {
+            observers[eventType] = [observer]
+        }
+    }
+
+    /// Removes all observers from the EventBus.
+    ///
+    /// This function is used for cleanup or resetting the event handling system.
+    func removeAllObservers() async {
+        observers.forEach { _, observerList in
+            observerList.forEach(notificationCenter.removeObserver)
+        }
+        observers.removeAll()
+    }
+
+    /// Removes all observers for a specific event type.
+    ///
+    /// - Parameter eventType: The event type for which to remove all observers.
+    func removeObserver(for eventType: String) async {
+        if let observerList = observers[eventType] {
+            for observer in observerList {
+                notificationCenter.removeObserver(observer)
+            }
+            observers[eventType] = nil
+        }
+    }
+}
```

Comment:
Due to the issue with removing the observers from the notification center on deinit, here is a suggestion.
In this one, we took out the observers dictionary and notification center in their own non-protected class (this need to be non-accessible outside the SharedEventBus), and we do the cleanup in its own deinit.
`EventBusObserversHolder` is not thread safe, hence it is only okay to use from the actor and should not expose it any further.

```suggestion
class EventBusObserversHolder {
  let notificationCenter: NotificationCenter = .default
  var observers: [String: [NSObjectProtocol]] = [:]
  
  func removeAllObservers() {
    observers.forEach { _, observerList in
      observerList.forEach(notificationCenter.removeObserver)
    }
    observers.removeAll()
  }
  
  deinit {
    self.removeAllObservers();
  }
}

actor SharedEventBus: EventBus {
  let holder = EventBusObserversHolder();

  init() {
    DIGraphShared.shared.logger.debug("SharedEventBus initialized")
  }
  
  /// Posts an event to all registered observers of its type.
  ///
  /// - Parameter event: The event to be posted.
  /// - Returns: True if the event has been posted to any observers, false otherwise.
  @discardableResult
  func post(_ event: AnyEventRepresentable) -> Bool {
    let key = event.key
    if let observerList = holder.observers[key], !observerList.isEmpty {
      holder.notificationCenter.post(name: NSNotification.Name(key), object: event)
      return true
    }
    return false
  }
  
  /// Registers an observer for a specific event type.
  ///
  /// - Parameters:
  ///   - eventType: The type of the event to observe.
  ///   - action: The action to be executed when the event is received.
  func addObserver(_ eventType: String, action: @escaping (AnyEventRepresentable) -> Void) {
    let observer = holder.notificationCenter.addObserver(forName: NSNotification.Name(eventType), object: nil, queue: nil) { notification in
      if let event = notification.object as? AnyEventRepresentable {
        action(event)
      }
    }
    // Store the observer reference for later management.
    if holder.observers[eventType] != nil {
      holder.observers[eventType]?.append(observer)
    } else {
      holder.observers[eventType] = [observer]
    }
  }
  
  /// Removes all observers from the EventBus.
  ///
  /// This function is used for cleanup or resetting the event handling system.
  func removeAllObservers() {
    holder.removeAllObservers();
  }
  
  /// Removes all observers for a specific event type.
  ///
  /// - Parameter eventType: The event type for which to remove all observers.
  func removeObserver(for eventType: String) {
    if let observerList = holder.observers[eventType] {
      for observer in observerList {
        holder.notificationCenter.removeObserver(observer)
      }
      holder.observers[eventType] = nil
    }
  }
}
```

Reply:
This can indeed be a good alternate, let me think about this.